### ISSUE TITLE
strdup() without POSIX?

### ISSUE THREAD

**Issue by gperciva** [Tarsnap/spiped#279]
`libcperciva/util/setuidgid.c` begins with
```
/* We use non-POSIX functionality in this file. */
#undef _POSIX_C_SOURCE
#undef _XOPEN_SOURCE
```
because it uses `setgroups()` [where possible].

Unfortunately, it also tries to use `strdup()`, which is a POSIX extension [1].  On FreeBSD, `strdup(3)` notes that
> The strdup() function is specified by IEEE Std 1003.1-2001 (“POSIX.1”).

while on Linux, I see:
> strdup():
>    _XOPEN_SOURCE >= 500
>       || /* Since glibc 2.12: */ _POSIX_C_SOURCE >= 200809L
>       || /* Glibc versions <= 2.19: */ _BSD_SOURCE || _SVID_SOURCE

[1] https://pubs.opengroup.org/onlinepubs/9699919799/functions/strdup.html

How should I proceed?

- my first instrict is to rewrite the code to use `malloc(3)`.
- should I try to play games with feature detection macros instead?

**Comment by gperciva** [comment]
Another option is to separate the posix and non-posix parts, i.e. add a `setuidgid_nonposix.c`.

**Comment by cperciva** [comment]
I'm not sure what the problem is here.  Sure, it's not in the C standard, but we're not expecting to run on any non-POSIX systems...

**Comment by gperciva** [comment]
`<string.h>` isn't required to define `stdup` if we haven't defined `_POSIX_C_SOURCE`.  And on one particular OS whose name begins with S, it doesn't.

**Comment by gperciva** [comment]
Fixed in #305.
